### DISCUSSION TITLE
Can I draw a rectangle?

### DISCUSSION THREAD

**Discussion by carecavoador** [pypdfium2-team/pypdfium2#160] [Q&A] [answered]
Hello,

I'd like to know if I can draw a rectangle on the page. I'm drawing a text on the page using the `insert_text` method, and it's fine. But, I wish I could draw a blank rectangle under it. Is it possible?

Edit: grammar.

**Comment by mara004** [comment, marked as answer]
There are no helper methods for it yet, but I think it can be done with the raw API (though I haven't tried this myself yet). `FPDFPageObj_CreateNewRect()` and `FPDFPageObj_SetFillColor()` should do the trick. Take a look at [`fpdf_edit.h`](https://pdfium.googlesource.com/pdfium/+/refs/heads/main/public/fpdf_edit.h).

Also note that the current `insert_text()` method has problems and will be removed or replaced with something simpler on the next major release. It currently attempts to insert text using an external text shaping engine (for foreign writing systems) but that is quite complex and the implementation has issues. As long as you don't need to support Hindi etc., you're better off with `FPDFPageObj_NewTextObj()` (or `FPDFPageObj_CreateTextObj()`) and `FPDFText_SetText()` until I have re-written the text insertion API.

**Reply by carecavoador** [reply]
Thank you! I'll check that out!

**Reply by mara004** [reply]
Info for readers of this thread: the old `insert_text()` helper was removed with pypdfium2 v4. A new, better text insertion helper might be re-implemented eventually.
